src/main/java/uk/gov/companieshouse/chipsrestinterfacesconsumer/configuration/KafkaConsumerConfig.java

```diff
@@ -0,0 +1,77 @@
+package uk.gov.companieshouse.chipsrestinterfacesconsumer.configuration;
+
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.kafka.annotation.EnableKafka;
+import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
+import org.springframework.kafka.core.ConsumerFactory;
+import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
+import uk.gov.companieshouse.chips.ChipsRestInterfacesSend;
+import uk.gov.companieshouse.chipsrestinterfacesconsumer.avro.AvroDeserializer;
+import uk.gov.companieshouse.chipsrestinterfacesconsumer.common.ApplicationLogger;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@EnableKafka
+@Configuration
+public class KafkaConsumerConfig {
+
+    @Value("${kafka.broker.address}")
+    private String brokerAddress;
+
+    @Value("${RETRY_THROTTLE_RATE_SECONDS}")
+    private long retryThrottleSeconds;
+
+    @Value("${MAX_RETRY_ATTEMPTS}")
+    private int maxRetryAttempts;
+
+    @Autowired
+    private ApplicationLogger logger;
+
+    @Bean
+    public ConsumerFactory<String, ChipsRestInterfacesSend> consumerFactory() {
+        Map<String, Object> props = new HashMap<>();
+        props.put(
+                ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
+                brokerAddress);
+        props.put(
+                ConsumerConfig.GROUP_ID_CONFIG,
+                "CRIC-Group");
+        props.put(
+                ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+                StringDeserializer.class);
+        props.put(
+                ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+                AvroDeserializer.class);
+        props.put(
+                ConsumerConfig.AUTO_OFFSET_RESET_CONFIG,
+                "earliest");
+        return new DefaultKafkaConsumerFactory<>(props, new StringDeserializer(), new AvroDeserializer<>(ChipsRestInterfacesSend.class));
+    }
+
+    private ConcurrentKafkaListenerContainerFactory<String, ChipsRestInterfacesSend> getMainFactory() {
+        ConcurrentKafkaListenerContainerFactory<String, ChipsRestInterfacesSend> factory =
+                new ConcurrentKafkaListenerContainerFactory<>();
+        factory.setConsumerFactory(consumerFactory());
+        return factory;
+    }
+
+    @Bean
+    public ConcurrentKafkaListenerContainerFactory<String, ChipsRestInterfacesSend>
```

Comment:
do we know what the default concurrency is for the factory? each concurrentMessageListenerContainer will create concurrent threads based on the concurrency value - https://docs.spring.io/spring-kafka/reference/html/#using-ConcurrentMessageListenerContainer. We should perhaps explicitly set each container concurrency for main and retry. We can have upto 1 consumer per partion on a topic. Might help with the concurrency requirments of the app if we can set the main and retry concurrency in config.

Reply:
basically, i'm thinking we might need to explicitly set the container.concurrency for the main and retry listener containers